client/network/src/protocol/notifications/behaviour.rs

```diff
@@ -2070,3 +2073,1985 @@ impl NetworkBehaviour for Notifications {
+		let peer = PeerId::random();
+		let conn = ConnectionId::new(0usize);
+		let set_id = sc_peerset::SetId::from(0);
+		let connected = ConnectedPoint::Listener {
+			local_addr: Multiaddr::empty(),
+			send_back_addr: Multiaddr::empty(),
+		};
+		let mut conn_yielder = ConnectionYielder::new();
+
+		// move the peer to `Enabled` state
+		notif.inject_connection_established(
+			&peer,
+			&conn,
+			&ConnectedPoint::Listener {
+				local_addr: Multiaddr::empty(),
+				send_back_addr: Multiaddr::empty(),
+			},
+			None,
+			0usize,
+		);
+		assert!(std::matches!(notif.peers.get(&(peer, set_id)), Some(&PeerState::Disabled { .. })));
+
+		notif.inject_event(peer, conn, NotifsHandlerOut::OpenDesiredByRemote { protocol_index: 0 });
+		assert!(std::matches!(notif.peers.get(&(peer, set_id)), Some(&PeerState::Incoming { .. })));
+
+		notif.peerset_report_connect(peer, set_id);
+		assert!(std::matches!(notif.peers.get(&(peer, set_id)), Some(&PeerState::Enabled { .. })));
+
+		let event = conn_yielder.open_substream(peer, 0, connected, vec![1, 2, 3, 4]);
+
+		notif.inject_event(peer, conn, event);
+		assert!(std::matches!(notif.peers.get(&(peer, set_id)), Some(&PeerState::Enabled { .. })));
+
+		if let Some(PeerState::Enabled { ref connections, .. }) = notif.peers.get(&(peer, set_id)) {
+			assert!(std::matches!(connections[0], (_, ConnectionState::Open(_))));
+			assert_eq!(connections[0].0, conn);
+		} else {
+			panic!("invalid state");
+		}
+
+		notif.peerset_report_disconnect(peer, set_id);
+
+		if let Some(PeerState::Disabled { ref connections, .. }) = notif.peers.get(&(peer, set_id))
+		{
+			assert!(std::matches!(connections[0], (_, ConnectionState::Closing)));
+			assert_eq!(connections[0].0, conn);
+		} else {
+			panic!("invalid state");
+		}
+
+		notif.peerset_report_disconnect(peer, set_id);
+
+		if let Some(PeerState::Disabled { ref connections, ref mut backoff_until }) =
+			notif.peers.get_mut(&(peer, set_id))
+		{
+			assert!(std::matches!(connections[0], (_, ConnectionState::Closing)));
+			assert_eq!(connections[0].0, conn);
+
+			*backoff_until =
+				Some(Instant::now().checked_add(std::time::Duration::from_secs(2)).unwrap());
+		} else {
+			panic!("invalid state");
+		}
+
+		notif.peerset_report_connect(peer, set_id);
+
+		let mut prev_instant =
+			if let Some(PeerState::DisabledPendingEnable {
+				ref connections, timer_deadline, ..
+			}) = notif.peers.get(&(peer, set_id))
+			{
+				assert!(std::matches!(connections[0], (_, ConnectionState::Closing)));
+				assert_eq!(connections[0].0, conn);
+
+				timer_deadline.clone()
+			} else {
+				panic!("invalid state");
+			};
+
+		if tokio::time::timeout(Duration::from_secs(5), async {
+			let mut params = MockPollParams { peer_id: PeerId::random(), addr: Multiaddr::empty() };
+
+			loop {
+				futures::future::poll_fn(|cx| {
+					let _ = notif.poll(cx, &mut params);
+					Poll::Ready(())
+				})
+				.await;
+
+				prev_instant =
```

Comment:
Why do we need to update `prev_instant` if after the first `break` the loop is terminated and it's not used anywhere after?

Reply:
Because the break is not necessarily triggered during the first iteration. I added a comment explaining how the test works. Basically it verifies that if the peer was `DisabledPendingEnable` but it doesn't have any valid connections (i.e., `Closed`), the code continues to keep the peer in that state until the connection is closed.

Reply:
Probably I'm missing something, but the code can proceed after the if
https://github.com/paritytech/substrate/blob/c1f59a355236b0d58f300a1b7d66ad73c3f550be/client/network/src/protocol/notifications/behaviour.rs#L3808-L3810
only if `timer_deadline == &prev_instant`, and in this case updating `prev_instant` to `timer_deadline` is a no-op.

Reply:
yes sorry, I see what you mean. I don't know why it was written like that, probably another copy-paste error